lib/OpenQA/Schema/Result/Jobs.pm

```diff
@@ -1051,35 +1051,55 @@ sub update_backend {
-    my ($self, $tm) = @_;
-
+    my ($self, $tm, $skip_jobs_update) = @_;
+
+    # prepare query to insert job module
+    my $insert_sth = $self->{_insert_job_module_sth};
+    $insert_sth = $self->{_insert_job_module_sth} = $self->result_source->schema->storage->dbh->prepare(
+        <<'END_SQL'
+        INSERT INTO job_modules (
+            job_id, name, category, script, milestone, important, fatal, always_rollback, t_created, t_updated
+        ) VALUES(
+            ?,      ?,    ?,        ?,      ?,         ?,         ?,     ?,               now(),      now()
+        ) ON CONFLICT DO NOTHING
+END_SQL
+    ) unless defined $insert_sth;
+
+    # execute query to insert job module
+    # note: We have 'important' in the DB but 'ignore_failure' in the flags for historical reasons (see #1266).
     my $flags = $tm->{flags};
-    try {
-        $self->modules->create(
-            {
-                name     => $tm->{name},
-                category => $tm->{category},
-                script   => $tm->{script},
-
-                # we have 'important' in the db but 'ignore_failure' in the flags
-                # for historical reasons...see #1266
-                milestone       => $flags->{milestone}       ? 1 : 0,
-                important       => $flags->{ignore_failure}  ? 0 : 1,
-                fatal           => $flags->{fatal}           ? 1 : 0,
-                always_rollback => $flags->{always_rollback} ? 1 : 0,
-            });
-    }
-    catch {
-        my $err = $_;
-        die $err unless $err =~ /job_modules_job_id_name_category_script/;
-    };
+    $insert_sth->execute(
+        $self->id, $tm->{name}, $tm->{category}, $tm->{script},
+        $flags->{milestone}       ? 1 : 0,
+        $flags->{ignore_failure}  ? 0 : 1,
+        $flags->{fatal}           ? 1 : 0,
+        $flags->{always_rollback} ? 1 : 0,
+    );
+    return 0 unless $insert_sth->rows;
+
+    # update job module statistics for that job (jobs with default result NONE are accounted as skipped)
+    $self->update({skipped_module_count => \'skipped_module_count + 1'}) unless $skip_jobs_update;
+    return 1;
 }
 
 sub insert_test_modules {
     my ($self, $testmodules) = @_;
-    for my $tm (@{$testmodules}) {
-        $self->insert_module($tm);
+    return undef unless scalar @$testmodules;
+
+    # insert all test modules and update job module statistics uxing txn to avoid inconsistent job module
+    # statistics in the error case
+    try {
+        $self->result_source->schema->txn_do(
+            sub {
+                my $new_rows = 0;
+                $new_rows += $self->insert_module($_, 1) for @$testmodules;
+                $self->update({skipped_module_count => \"skipped_module_count + $new_rows"});
+            });
     }
+    catch {
+        my $job_id = $self->id;                                            # uncoverable statement
+        log_warning("Failed to insert job modules of job $job_id: $_");    # uncoverable statement
+    };
```

Comment:
and why would we need a try/catch here? Wouldn't this also cause errors in postgres log? As you also added "# uncoverable statement" I assume you did not manage to cover this by tests so I wonder why this should be necessary then

Reply:
I added the transaction because I want to avoid inconsistent database updates when an unexpected error occurs in the middle. However, there's indeed no error to be expected. (The database connection might tear down in the middle but that kind of error we usually never explicitly handle.) So I'm removing the try-catch-block.